typing/shape_reduce.ml

```diff
@@ -323,7 +331,7 @@ end) = struct
       Unresolved (read_back env nf)
```

Comment:
This occurrence of `read_back` is also a risk for performance, right? (Did we discuss making this a `unit -> shape` function to let the user force the read-back on demand?)

Reply:
You are right that there is no obviously good choice here. I am ok with forgetting about that case and focusing on the other one, that you identified as being an issue in practice, for the present PR.